watcher/src/block_data_store.rs

```diff
@@ -19,44 +22,54 @@ use url::Url;
 /// Block datas database name.
 pub const BLOCK_DATAS_BY_INDEX_DB_NAME: &str = "watcher_db:block_data:blocks_datas_by_index";
 
-/// Blocks by hash database name.
+/// [Block]s by hash database name.
 pub const BLOCKS_BY_HASH_DB_NAME: &str = "watcher_db:block_data:blocks_by_hash";
 
-/// BlockContents by hash database name.
+/// [BlockContents] by hash database name.
 pub const BLOCK_CONTENTS_BY_HASH_DB_NAME: &str = "watcher_db:block_data:block_contents_by_hash";
 
+/// [BlockMetadata] by hash database name.
+pub const BLOCK_METADATA_BY_HASH_DB_NAME: &str = "watcher_db:block_data:block_metadata_by_hash";
+
 /// An internal object for representing BlockData that doesn't hold the actual
 /// Block and BlockContents since those might be shared with other blocks.
 #[derive(Clone, Message)]
 pub struct StoredBlockData {
-    /// 32 bytes hash of Block.
-    #[prost(bytes, required, tag = "1")]
+    /// 32 bytes hash of [Block].
+    #[prost(bytes, required, tag = 1)]
     pub block_hash: Vec<u8>,
 
-    /// 32 bytes hash of BlockContent.
-    #[prost(bytes, required, tag = "2")]
+    /// 32 bytes hash of [BlockContents].
+    #[prost(bytes, required, tag = 2)]
     pub block_contents_hash: Vec<u8>,
 
     /// Block signature (optional).
     // The signature is unique (we do not expect to encounter duplicate signatures)
     // so we store it inside here.
-    #[prost(message, tag = "3")]
+    #[prost(message, tag = 3)]
     pub signature: Option<BlockSignature>,
+
+    /// Hash of [BlockMetadata].
+    #[prost(bytes, optional, tag = 4)]
+    pub block_metadata_hash: Option<Vec<u8>>,
```

Comment:
I missed this in previous reviews, but I don't think this should be storing the metadata hash, similarly to how it doesn't store the signature hash.
The idea of storing hashes of the `Block` and `BlockContents` is that they are expected to be identical amongst all instances of a given block, but the signatures will vary for each source - or to phrase differently, for each node we "watch" we will be getting a list of `StoredBlockData` for the blocks produced by that node. For a given block index, the Block and BlockContents will very very likely be the same, so we avoid storing them multiple times in the database and instead store the hash here, and have a separate table to map hash -> actual contents.
We don't do that for the signature since that is expected to be unique, and in most cases the metadata will also be unique - so it should be stored here as `Option<BlockMetadata>`. This will likely simplify the changes here since there won't be a need for the new `BLOCK_METADATA_BY_HASH_DB_NAME` database.
Apologies for not catching it earlier.

Reply:
Good catch!

> Apologies for not catching it earlier.

No worries; that's why I pared down this PR.